metadata_service/proxy/atlas_proxy.py

```diff
@@ -652,6 +689,66 @@ def _parse_date(self, date: int) -> Optional[int]:
         except Exception:
             return None
 
+    @staticmethod
+    def _extract_entities(collections: EntityCollection) -> List[Entity]:
+        """
+        Helper method for flattening all collections from {collections}
+
+        :return: list of all entities
+        """
+        entities: List[Entity] = []
+
+        for collection in collections:
+            entities.extend(collection.entities)
+        return entities
+
+    def _get_readers(self, qualified_name: str, top: Optional[int] = None) -> List[Reader]:
+        params = {
+            'typeName': self.READER_TYPE,
+            'offset': '0',
+            'limit': top or 15,
+            'excludeDeletedEntities': True,
+            'entityFilters': {
+                'condition': 'AND',
+                'criterion': [
+                    {
+                        'attributeName': self.QN_KEY,
+                        'operator': 'STARTSWITH',
+                        'attributeValue': qualified_name.split('@')[0]
+                    },
+                    {
+                        'attributeName': self.READER_COUNT_KEY,
+                        'operator': 'gt',
+                        'attributeValue': '0'
+                    }
+                ]
+            },
+            'attributes': [self.READER_COUNT_KEY, self.QN_KEY],
+            'sortBy': self.READER_COUNT_KEY,
+            'sortOrder': 'DESCENDING'
+        }
+
+        search_results = self._driver.search_basic.create(data=params, ignoreRelationships=False)
+
+        readers = []
+
+        for record in search_results.entities:
+            readers.append(record.guid)
+
+        results = []
+
+        if len(readers) > 0:
+            full_entities = AtlasProxy._extract_entities(self._driver.entity_bulk(guid=readers,
+                                                                                  ignoreRelationships=False))
+
+            for e in full_entities:
+                r = Reader(user=User(email=e.relationshipAttributes['user']['displayText']),
```

Comment:
Please fill in the `user_id` field instead. We'll be making the email as optional and user_id as a required attribute. and also, the whole communication of frontend goes through `user_id`.

Reply:
that was first implementation but turns out reader schema expects email, not user_id:

` "Encountered exception: {'table_readers': {0: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 1: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 2: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 3: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 4: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 5: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 6: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}, 7: {'user': {'_schema': ['"display_name", "full_name", or "email" must be provided']}}}}"`

Reply:
oh yes, so you could fill in user_id and display_name/email with the CK. Will have a custom configurable function to get the user details wherever needed in the future. Main point is to fill in the user_id here.

Reply:
done